src/Dhall/Core.hs

```diff
@@ -1130,6 +1142,7 @@ normalize e = case e of
             App NaturalIsZero (NaturalLit n) -> BoolLit (n == 0)
             App NaturalEven (NaturalLit n) -> BoolLit (even n)
             App NaturalOdd (NaturalLit n) -> BoolLit (odd n)
+            -- What does this do? (TODO)
```

Comment:
So what is going on here is that the user supplies a function to build the list which might look like this:

```haskell
List/build
Integer
(   λ(list : Type)
→   λ(cons : Integer → list → list)
→   λ(nil  : list)
→   cons 1 (cons 2 (cons 3 nil))
)
```

... and the abstract syntax tree for that would be:

```haskell
App
(App ListBuild Integer)
(Lam "list" Type
    (Lam "cons" (Pi "_" Integer (Pi "_" "list" "list"))
        (Lam "nil" "list"
            (App (App "cons" 1) (App (App "cons" 2) (App (App "cons" 3) "nil"))) ) ) )
```

... and we basically need to turn that syntax tree into a vector.

One simple approach to do this would be to pattern match on the syntax tree, extract the labels that the user chose for `cons` and `nil` and use those labels to detect if the built list is correct.  The code for that might look something like this:

```haskell
k' cons nil = go0 k

go0 (App _ (Lam l _ (Lam c _ (Lam n _ e)))) = go1 e
  where
    go (App (App (Var (V c' 0)) x) e')
        | c == c' = cons x (go e')
    go (Var (V n' 0))
        | n == n' = nil
    go _ = error "Urk!"
```

... however that's not exactly right because it doesn't deal with variable shadowing.  For example, it gives the wrong answer if the user supplies the same variable name for the `cons` and `nil` arguments, like this:

```haskell
List/build
Integer
(   λ(list : Type)
→   λ(x : Integer → list → list)
→   λ(x  : list)
→   x@1 1 (x@1 2 (x@1 3 x))
)
```

The way I worked around this was to use `dhall`'s normalize functionality to replace the user's chosen label names with labels of my choice.  In other words, I take the user's function and apply it to two unique labels like this:

```haskell
(   λ(list : Type)
→   λ(x : Integer → list → list)
→   λ(x  : list)
→   x@1 1 (x@1 2 (x@1 3 x))
)
(List t)
Cons
Nil
```

... which reduces to:

```haskell
Cons 1 (Cons 2 (Cons 3 Nil))
```

That ensures that the variable names are unique so that I don't have to worry about shadowing.

This is not the most efficient implementation, but it was the easiest to implement at the time.